### PR TITLE
Add cert-manager extension support

### PR DESCRIPTION
**Pull request by matthyx**
## Type of change

 - [ ] Bug fix (non-breaking change which fixes an issue)
 - [x] Feature (non-breaking change which adds functionality)
 - [ ] Breaking change (fix or feature that would cause existing functionality to change
 - [ ] Chore (non-breaking change which doesn't affect codebase;
   test, version modification, documentation, etc.)

## Checklist
 - [x] Code contributed by me aligns with current project license: [Apache 2.0](https://www.apache.org/licenses/LICENSE-2.0)
 - [ ] I Added [CHANGELOG](../CHANGELOG.md) entry regarding this change
 - [ ] I have implemented unit tests to cover my changes
 - [ ] I have added/updated the [javadocs](https://www.javadoc.io/doc/io.fabric8/kubernetes-client/latest/index.html) and other [documentation](https://github.com/fabric8io/kubernetes-client/blob/master/doc/CHEATSHEET.md) accordingly
 - [ ] No new bugs, code smells, etc. in [SonarCloud](https://sonarcloud.io/dashboard?id=fabric8io_kubernetes-client) report
 - [ ] I tested my code in Kubernetes
 - [ ] I tested my code in OpenShift

### CODE REVIEW

**Comment by matthyx** [comment]
relates to https://github.com/fabric8io/kubernetes-client/issues/2565

**Comment by matthyx** [comment]
@rohanKanojia here is where I am now:

- go generator works
- I have java files inside `extensions/certmanager/model/target/generated-sources`
- unfortunately nothing ends up in any of the `src` folders

What am I doing wrong?

**Comment by rohanKanojia** [comment]
let me try to clone your branch and check later tonight

**Comment by matthyx** [comment]
> let me try to clone your branch and check later tonight

sure, no rush... I am not the best Maven specialist, so it's probably down to my limited understanding of the whole mechanism.

**Comment by matthyx** [comment]
Wow, diff is epic and too much to check on the phone... I'll see later with the laptop while catching up on Netflix with my wife ;-)

**Comment by matthyx** [comment]
So basically you started from an older version of master, still 5.2?
The change that was needed for the java files to appear was to delete an ant task, or did you copy by hand?

File: extensions/certmanager/generator/cmd/generate/generate.go [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+	// e.g. knative.dev/eventing/pkg/apis/messaging/v1beta1/ChannelTemplateSpec is mapped to "io.fabric8.knative.internal.eventing.pkg.apis.messaging.v1beta1.ChannelTemplateSpec"
+	mappingSchema := map[string]string{
+	}
+
+	// overwriting some times
+	manualTypeMap := map[reflect.Type]string{
+		reflect.TypeOf(v1.Time{}):                  "java.lang.String",
+		reflect.TypeOf(metav1.ObjectReference{}):   "java.lang.String",
+		reflect.TypeOf(metav1.SecretKeySelector{}): "java.lang.String",
+                reflect.TypeOf(apiextensions.JSON{}): "com.fasterxml.jackson.databind.JsonNode",
```
Comment (matthyx) [inline extensions/certmanager/generator/cmd/generate/generate.go:77]: not sure gofmt is happy with that...

File: extensions/certmanager/generator/cmd/generate/generate.go [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+	"reflect"
+)
+
+func main() {
+
+	// the CRD List types for which the model should be generated
+	// no other types need to be defined as they are auto discovered
+	crdLists := map[reflect.Type]schemagen.CrdScope{
+		reflect.TypeOf(certmanager.CertificateList{}): schemagen.Namespaced,
+		reflect.TypeOf(certmanager.CertificateRequestList{}): schemagen.Namespaced,
```
Comment (matthyx) [inline extensions/certmanager/generator/cmd/generate/generate.go:35]: how do you know which types to add? gut feeling?

**Review by matthyx — COMMENTED** [review]
thanks for fixing my branch @rohanKanojia I just commented few places to understand the process
should I rebase on 5.3-SNAPSHOT?

File: extensions/certmanager/generator/cmd/generate/generate.go (hunk 2 above)
Comment (rohanKanojia) [reply extensions/certmanager/generator/cmd/generate/generate.go:35]: I checked their provided CRDs: https://github.com/jetstack/cert-manager/tree/master/deploy/crds and added whatever I could

**Comment by rohanKanojia** [comment]
I tested building project on jdk11 before pushing to your branch. Quite strange it's failing on jdk8 due to some files getting generated twice. Let me add antrun tasks to delete duplicates.

**Comment by rohanKanojia** [comment]
> So basically you started from an older version of master, still 5.2?
The change that was needed for the java files to appear was to delete an ant task, or did you copy by hand?

Umm, I think I did rebase your branch with master. I remember I replaced 5.2-SNAPSHOT with 5.3-SNAPSHOT. I had to copy `ACMEChallengeSolverHTTP01Ingress` to `src/` since it has a[ field name `Class`](https://github.com/jetstack/cert-manager/blob/master/pkg/apis/acme/v1/types_issuer.go#L216) which was causing build failures during sundrio generation

**Comment by matthyx** [comment]
> > So basically you started from an older version of master, still 5.2?
> > The change that was needed for the java files to appear was to delete an ant task, or did you copy by hand?
> 
> Umm, I think I did rebase your branch with master. I remember I replaced 5.2-SNAPSHOT with 5.3-SNAPSHOT. I had to copy `ACMEChallengeSolverHTTP01Ingress` to `src/` since it has a[ field name `Class`](https://github.com/jetstack/cert-manager/blob/master/pkg/apis/acme/v1/types_issuer.go#L216) which was causing build failures during sundrio generation

thanks for the explanation, everyday learning!

**Comment by rohanKanojia** [comment]
Bdw, [cert-manager ](https://github.com/jetstack/cert-manager/tree/master/pkg/apis/acme)seems to have models for `v1`, `v1alpha2`, `v1alpha3`, `v1beta1` . Shall we add those models too? Do you know if cert-manager recently graduated it's api to `v1` and whether it's going to stay there for long term?

**Comment by matthyx** [comment]
> Bdw, [cert-manager ](https://github.com/jetstack/cert-manager/tree/master/pkg/apis/acme)seems to have models for `v1`, `v1alpha2`, `v1alpha3`, `v1beta1` . Shall we add those models too(right now we only have `v1`)? Do you know if cert-manager recently graduated it's api to `v1` and whether it's going to stay there for long term?

I have no idea... but I never saw anything else than v1 in the examples. I'll keep looking and worse case I ask them.

**Comment by matthyx** [comment]
So, @munnerz replied on slack that v1 is supported since cert-manager 1.0.0 (released sept 2020) but older ones are still considered "valid".
If we want to help people on older installs (OpenShift 3.11 anyone?) we should support all of them...

Now that I see how you did it, I can work on that tomorrow if you want :-)

**Comment by matthyx** [comment]
@rohanKanojia I don't know about the CI errors here, but I think it only remains to add some examples and tests...

**Comment by matthyx** [comment]
@rohanKanojia I need your help here... impossible to understand how the generated sources are supposed to move from target/ to somewhere in src/
I think I have all good regarding the generators, but somehow missing some parts on the Maven side. Do you have some time to help me on that, please?

**Comment by rohanKanojia** [comment]
@matthyx : Sorry for the late reply. I don't think you need to worry about moving sources from `target/` to `src/` folder. In Kubernetes Model, most of our sources are being generated at compile time and they are not persisted into version control.

**Comment by rohanKanojia** [comment]
@matthyx : Hi, Are you still working on this? If you think it's ready could you please remove WIP from the PR title? I can request people for review then

**Comment by matthyx** [comment]
Hi Rohan, sorry it's not ready. I will work on this next week I promise :-)

**Comment by rohanKanojia** [comment]
No worries, no need to rush. Take your time.

**Comment by matthyx** [comment]
@rohanKanojia I have added the various api versions, however I get some build failures... could you take a look, I don't understand why it fails to find the generated classes.

**Comment by rohanKanojia** [comment]
@matthyx : I think we would also need to add ExtensionAdapters for each client. Something like this: https://github.com/fabric8io/kubernetes-client/blob/master/extensions/tekton/client/src/main/java/io/fabric8/tekton/client/V1alpha1APIGroupExtensionAdapter.java

**Comment by matthyx** [comment]
> @matthyx : I think we would also need to add ExtensionAdapters for each client. Something like this: https://github.com/fabric8io/kubernetes-client/blob/master/extensions/tekton/client/src/main/java/io/fabric8/tekton/client/V1alpha1APIGroupExtensionAdapter.java

Sure, thanks for catching the duplicate property file... this work is a bit overwhelming to be honest :)

**Comment by rohanKanojia** [comment]
ohk, It's very close to completion. We just need to write extension adapters and write some tests. I can take over from here if you don't have time.

**Comment by matthyx** [comment]
> ohk, It's very close to completion. We just need to write extension adapters and write some tests. I can take over from here and complete it on weekend if you don't have time.

I have added the adapters locally, and now I am going to add some more examples and yes tests too... it's easier to work on them when IDEA no longer screams with errors :)

**Comment by matthyx** [comment]
> > ohk, It's very close to completion. We just need to write extension adapters and write some tests. I can take over from here and complete it on weekend if you don't have time.
> 
> I have added the adapters locally, and now I am going to add some more examples and yes tests too... it's easier to work on them when IDEA no longer screams with errors :)

Thanks a lot for your precious help!

**Comment by rohanKanojia** [comment]
np, thanks a lot for your work :clap:

**Comment by matthyx** [comment]
@rohanKanojia I don't know if you want/need more examples and tests?
Also what to do with the sonarcloud result?

**Comment by rohanKanojia** [comment]
Regarding [SonarCloud warning about gpg passhrase](https://sonarcloud.io/project/security_hotspots?id=fabric8io_kubernetes-client&pullRequest=2930&resolved=false&types=SECURITY_HOTSPOT) in pom. I think we can remove gpg-plugin from cert-manager pom as it will be picked up by parent pom anyway. I'll try to refactor it later to pick it from some environment variable rather than hardcoding in pom.xml

Don't worry about code smells. You are free to fix which you think are valid.

It would be nice to add some `load()` tests to check whether our model is deserializing Cert Manager resources properly

**Comment by matthyx** [comment]
> It would be nice to add some `load()` tests to check whether our model is deserializing Cert Manager resources properly

done

**Comment by rohanKanojia** [comment]
cool, Thanks. Do you think it's ready for review? Shall I request for review?

**Comment by matthyx** [comment]
yes please!

File: extensions/certmanager/client/src/main/java/io/fabric8/certmanager/client/CertManagerExtensionAdapter.java [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+import java.net.URL;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+
+public class CertManagerExtensionAdapter extends ExtensionAdapterSupport implements ExtensionAdapter<CertManagerClient> {
+
+  static final ConcurrentMap<URL, Boolean> IS_VOLUME_SNAPSHOT = new ConcurrentHashMap<>();
+  static final ConcurrentMap<URL, Boolean> USES_VOLUME_SNAPSHOT_APIGROUPS = new ConcurrentHashMap<>();
+  public static final String API_GROUP = "snapshot.storage.k8s.io";
```
Comment (rohanKanojia) [inline extensions/certmanager/client/src/main/java/io/fabric8/certmanager/client/CertManagerExtensionAdapter.java:31]: I think these should be renamed as per this extension's name/apigroup

Comment (matthyx) [reply extensions/certmanager/client/src/main/java/io/fabric8/certmanager/client/CertManagerExtensionAdapter.java]: indeed, thanks!

**Review by rohanKanojia — APPROVED** [review]

**Comment by rohanKanojia** [comment]
Thanks a lot for this :clap:

**Comment by vsevel** [comment]
good job @matthyx

**Comment by matthyx** [comment]
@manusa do you know for which release this is planned, please?

**Comment by manusa** [comment]
> @manusa do you know for which release this is planned, please?

Next minor release should be fine.

I tried to rebase your changes, but it required some additional work. I'm not sure if you can take care of this. If not, I'll try to do that next week.

**Comment by matthyx** [comment]
> I tried to rebase your changes, but it required some additional work. I'm not sure if you can take care of this. If not, I'll try to do that next week.

Let me have a try...

**Comment by irbekrm** [comment]
> Bdw, cert-manager seems to have models for v1, v1alpha2, v1alpha3, v1beta1 . Shall we add those models too(right now we only have v1)? Do you know if cert-manager recently graduated it's api to v1 and whether it's going to stay there for long term?

Hi folks 👋🏼 A drive-by comment that might or might not be relevant for you- the `alpha` and `beta` APIs are getting deprecated in the next release of `cert-manager` (`v1.4`) and will be removed in `v1.6` cert-manager#4021](https://github.com/jetstack/cert-manager/pull/4021). (Also the `v1` API _is_ considered stable and long-term).

**Comment by matthyx** [comment]
awesome, maybe it will help me make my tests pass :-)

@manusa I am still failing to fix them, as for some reason my IntelliJ fails to compile the current project at master

**Comment by manusa** [comment]
> @manusa I am still failing to fix them, as for some reason my IntelliJ fails to compile the current project at master

Could you provide more details about you failure?
Have you tried delegating the IntelliJ build to Maven?

**Comment by matthyx** [comment]
again those dreadful:
```
java: java.lang.NoClassDefFoundError: io/fabric8/kubernetes/client/extended/run/RunConfig
io.fabric8.kubernetes.client.extended.run.RunConfig
```

**Comment by manusa** [comment]
What about the Maven build delegation?
image

**Comment by matthyx** [comment]
> What about the Maven build delegation?
> image

Thanks, trying... (I swear I didn't need it in the past)

**Comment by matthyx** [comment]
@manusa could you check why the Karaf test failed? I swear I didn't touch it!

**Comment by manusa** [comment]
> @manusa could you check why the Karaf test failed? I swear I didn't touch it!

Those are flaky. We'll need to look into them, but for some reason there are timeouts when starting the platform.

**Comment by matthyx** [comment]
> > @manusa could you check why the Karaf test failed? I swear I didn't touch it!
> 
> Those are flaky. We'll need to look into them, but for some reason there are timeouts when starting the platform.

So I think we should merge this PR quickly until we need another headache rebase...

**Review by manusa — APPROVED** [review]
LGTM, thx!

**Pull request merged**